src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs

```diff
@@ -506,6 +506,42 @@ private TypeSpec CreateEnumerableSpec(TypeParseInfo typeParseInfo)
                 };
             }
 
+            private bool ConvertTo(ITypeSymbol source, ITypeSymbol dest)
+            {
+                var conversion = _typeSymbols.Compilation.ClassifyConversion(source, dest);
+                return conversion.IsReference && conversion.IsImplicit;
+            }
+
+            private bool IsUnsupportedType(ITypeSymbol type)
+            {
+                if (SymbolEqualityComparer.Default.Equals(_typeSymbols.IntPtr, type)  ||
+                    SymbolEqualityComparer.Default.Equals(_typeSymbols.UIntPtr, type) ||
+                    SymbolEqualityComparer.Default.Equals(_typeSymbols.SerializationInfo, type) ||
+                    ConvertTo(type, _typeSymbols.MemberInfo) ||
+                    ConvertTo(type, _typeSymbols.Delegate))
+                {
+                    return true;
+                }
+
+                if (type is IArrayTypeSymbol arrayTypeSymbol)
+                {
+                    return arrayTypeSymbol.Rank > 1 || IsUnsupportedType(arrayTypeSymbol.ElementType);
+                }
+
+                if (type is INamedTypeSymbol namedTypeSymbol && namedTypeSymbol.IsGenericType)
+                {
+                    foreach (var typeArguments in namedTypeSymbol.TypeArguments)
```

Comment:
I don't think walking the type arguments and failing if one of them is unsupported is the correct approach. For example, I could have something like:
```C#
public record MyPoco<TUnused>(int Value);
```
Then binding configuration for something like `MyPoco<Func<int>>` would be perfrectly valid from data contract perspective. I think we should only be rejecting types if one of their property types (or element types in the case of collections) is unsupported.

Reply:
We thought about that and preferred to keep it like that. The reason is we already have a real case doing that and we shouldn't generate code for, in the original issue I pointed at example https://github.com/dotnet/extensions/blob/14943b99572078f1bcf2c51e9507cf2631133fae/src/Libraries/Microsoft.Extensions.Resilience/Resilience/Internal/FailureEventMetricsOptions.cs#L11. Second, if we run into some issue in the future (as this is not a popular scenario) we can consider tweaking this if needed. For now, we should be a little bit aggressive.

Reply:
That particular example though falls under the element type not being supported. You would still need to check for that independently because it may or may not appear in the generic arguments of the particular type, i.e.
```C#
internal sealed class FailureEventMetricsOptions
{
    // No type arguments visible in the property type.
    public MyCustomDictionary  Factories { get; } = [];
}

public class MyCustomDictionary : IDictionary<Dictionary<Type, Func<object, FailureResultContext>>>
{
}
```
Point being, type arguments are independent of what data the type contains.

Reply:
I applied your suggestion not checking the generic type parameters and instead I am covering the collection types.